[GStreamer][MSE] Optional encoders for AppendPIpeline streams

vivienne-w opened the pull request:
#### 7926de4222132a135d84b4d4cc05d8f6738f9c6a
<pre>
[GStreamer][MSE] Optional encoders for AppendPIpeline streams
<a href="https://bugs.webkit.org/show_bug.cgi?id=287860">https://bugs.webkit.org/show_bug.cgi?id=287860</a>

Reviewed by Philippe Normand.

Adds optional encoder elements, analogous to the optional parsers streams can have.
This is necessary for TextTrack support in MSE, where we need to re-encode any
subtitle or caption tracks to WebVTT, which is the format WebKit expects.

A stream in AppendPipeline now looks like this:
               __________    ___________    ___________
___________    |        |    |         |    |         |
|         | -&gt; | parser | -&gt; | encoder | -&gt; | appsink |
| demuxer |    |________|    |_________|    |_________|
|_________| -&gt; ...

Additionally, I&apos;ve renamed parser elements, which previously tended to break pipeline dumps.
See: <a href="https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/7979">https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/7979</a>
While this was backported to 1.24.10, the change also makes element names more descriptive.
(Take for example 0_parser, which might now be called parser_0_video.)

Finally, I&apos;ve changed some logging in TrackPrivateBaseGStreamer, which didn&apos;t yet use track IDs.

* Source/WebCore/platform/graphics/gstreamer/TrackPrivateBaseGStreamer.cpp:
(WebCore::TrackPrivateBaseGStreamer::tagsChanged):
(WebCore::TrackPrivateBaseGStreamer::getLanguageCode):
(WebCore::TrackPrivateBaseGStreamer::getTag):
(WebCore::TrackPrivateBaseGStreamer::notifyTrackOfStreamChanged):
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
(WebCore::AppendPipeline::appsinkCapsChanged):
(WebCore::AppendPipeline::didReceiveInitializationSegment):
(WebCore::createOptionalParserForFormat):
  Just takes the new element&apos;s name as argument now.
(WebCore::createOptionalEncoderForFormat):
  Creates an encoder, analogous to its parser counterpart.
(WebCore::AppendPipeline::recycleTrackForPad):
(WebCore::AppendPipeline::makeWebKitTrack):
(WebCore::AppendPipeline::Track::emplaceOptionalParserForFormat):
  Creates parserName, and links the parser to the encoder, instead of the appsink.
(WebCore::AppendPipeline::Track::emplaceOptionalEncoderForFormat):
  Create an encoder and link it to the appsink, also set finalCaps here.
(WebCore::AppendPipeline::Track::initializeElements):
(WebCore::AppendPipeline::streamTypeToStringLower):
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.h:
  Add fields to AppendPipeline::Track: finalCaps, encoder, and encoderPad.

Canonical link: <a href="https://commits.webkit.org/290605@main">https://commits.webkit.org/290605@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f1a41c9d5829cd48b23265410ff66f5b3ca4683

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90432 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9962 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45360 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95442 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41214 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92484 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10355 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18283 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69617 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27186 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93433 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7933 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82034 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49966 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7655 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36386 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40344 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77990 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37470 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97270 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17624 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12959 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78606 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17882 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77853 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77826 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19243 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22272 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20886 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/10946 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17634 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22966 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17373 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20828 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19157 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -942,19 +963,38 @@ void AppendPipeline::Track::emplaceOptionalParserForFormat(GstBin* bin, const GR
 
+void AppendPipeline::Track::emplaceOptionalEncoderForFormat(GstBin* bin, const GRefPtr<GstCaps>& newCaps)
+{
+    auto encoderName = makeString("encoder"_s, String::fromLatin1(streamTypeToStringLower(streamType)), "_"_s, trackId);
```
i think `span8()` should be usable here instead of `String::fromLatin1`

philn commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -769,6 +769,26 @@ GRefPtr<GstElement> createOptionalParserForFormat([[maybe_unused]] GstBin* bin,
     return result;
 }
 
+GRefPtr<GstElement> createOptionalEncoderForFormat([[maybe_unused]] GstBin* bin, String encoderName, const GstCaps* caps)
```
`const String&`

philn commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -684,7 +685,7 @@ void AppendPipeline::handleAppsinkNewSampleFromStreamingThread(GstElement*)
 }
 
-GRefPtr<GstElement> createOptionalParserForFormat([[maybe_unused]] GstBin* bin, TrackID trackId, const GstCaps* caps)
+GRefPtr<GstElement> createOptionalParserForFormat([[maybe_unused]] GstBin* bin, String parserName, const GstCaps* caps)
```
`const String&` ;)

vivienne-w committed:
[GStreamer][MSE] Optional encoders for AppendPIpeline streams
https://bugs.webkit.org/show_bug.cgi?id=287860

Reviewed by Philippe Normand.

Adds optional encoder elements, analogous to the optional parsers streams can have.
This is necessary for TextTrack support in MSE, where we need to re-encode any
subtitle or caption tracks to WebVTT, which is the format WebKit expects.

A stream in AppendPipeline now looks like this:
               __________    ___________    ___________
___________    |        |    |         |    |         |
|         | -> | parser | -> | encoder | -> | appsink |
| demuxer |    |________|    |_________|    |_________|
|_________| -> ...

Additionally, I've renamed parser elements, which previously tended to break pipeline dumps.
See: https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/7979
While this was backported to 1.24.10, the change also makes element names more descriptive.
(Take for example 0_parser, which might now be called parser_0_video.)

Finally, I've changed some logging in TrackPrivateBaseGStreamer, which didn't yet use track IDs.

* Source/WebCore/platform/graphics/gstreamer/TrackPrivateBaseGStreamer.cpp:
(WebCore::TrackPrivateBaseGStreamer::tagsChanged):
(WebCore::TrackPrivateBaseGStreamer::getLanguageCode):
(WebCore::TrackPrivateBaseGStreamer::getTag):
(WebCore::TrackPrivateBaseGStreamer::notifyTrackOfStreamChanged):
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
(WebCore::AppendPipeline::appsinkCapsChanged):
(WebCore::AppendPipeline::didReceiveInitializationSegment):
(WebCore::createOptionalParserForFormat):
  Just takes the new element's name as argument now.
(WebCore::createOptionalEncoderForFormat):
  Creates an encoder, analogous to its parser counterpart.
(WebCore::AppendPipeline::recycleTrackForPad):
(WebCore::AppendPipeline::makeWebKitTrack):
(WebCore::AppendPipeline::Track::emplaceOptionalParserForFormat):
  Creates parserName, and links the parser to the encoder, instead of the appsink.
(WebCore::AppendPipeline::Track::emplaceOptionalEncoderForFormat):
  Create an encoder and link it to the appsink, also set finalCaps here.
(WebCore::AppendPipeline::Track::initializeElements):
(WebCore::AppendPipeline::streamTypeToStringLower):
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.h:
  Add fields to AppendPipeline::Track: finalCaps, encoder, and encoderPad.

Canonical link: https://commits.webkit.org/290605@main
